src/Nethermind/Nethermind.Evm/VirtualMachine.cs

```diff
@@ -1545,6 +1545,31 @@ static void UpdateCurrentState(EvmState state, int pc, long gas, int stackHead)
                         vmState.Memory.SaveByte(in memPosition, data);
                         if (_txTracer.IsTracingInstructions) _txTracer.ReportMemoryChange((long)memPosition, data);
 
+                        break;
+                    }
+                case Instruction.MCOPY:
+                    {
+                        if (!spec.MCopyIncluded)
+                        {
+                            goto InvalidInstruction;
+                        }
+
+                        Metrics.MCopyOpcode++;
+
+                        stack.PopUInt256(out UInt256 dstPosition);
+                        stack.PopUInt256(out UInt256 srcPosition);
+                        stack.PopUInt256(out UInt256 length);
+
+                        long baseGasCost = GasCostOf.VeryLow + GasCostOf.VeryLow * (((long)length + 31) / 32);
+                        if (!UpdateGas(baseGasCost, ref gasAvailable)
+                            || !UpdateMemoryCost(vmState, ref gasAvailable, UInt256.Max(srcPosition, dstPosition), length)) goto OutOfGas;
+
+                        Span<byte> loadedData = vmState.Memory.LoadSpan(in srcPosition, length);
+                        if (_txTracer.IsTracingInstructions) _txTracer.ReportMemoryChange(srcPosition, loadedData);
+
+                        vmState.Memory.Save(in dstPosition, loadedData);
```

Comment:
This should handle overlapping memory just fine, right?

Reply:
I wrote a test case for this scenario and it passes, if you have lore complex scenarios to test please do tell

Reply:
It calls `Span<byte>.CopyTo` which calls into `Buffer.Memmove` which has a check and handling for overlapping buffers in the runtime https://github.com/dotnet/runtime/blob/8770047887d05d106699badd27520a7b260a99be/src/libraries/System.Private.CoreLib/src/System/Buffer.cs#L131-L137

Reply:
Yeah I thought so, just double-checking :)